doc/compiler/jitserver/Caching.md

```diff
@@ -19,35 +19,75 @@ OpenJDK Assembly Exception [2].
 -->
+
 # Caching in JITServer
-Caching is an important aspect of `JITServer` and one of the main reasons for why `JITServer` client consumes less CPU time compared to local compilation. This document will explain why caching is important, what kind of entities we cache, and how some specific caches work.
 
-## Why caching is important
-When server is compiling a method, it needs to know a lot of information that is located on the client side, e.g. the addresses of RAM classes, GC mode, class hierarchy information, interpreter profiling data and much more. To obtain this information, server will make a remote call over the network to the client, client will find it, and send it back to the server. During the course of a compilation server will make hundreds of such calls (on average), assuming no caching is involved.
-Why is it bad? First of all, remote calls are expensive, both in terms of latency and CPU, especially in terms of CPU. If the server makes hundreds of requests to the client just for one compilation, client will spend more CPU time just sending and receiving data over the network than it would take for it to compile a method locally. What makes it worse is that in order for the client to fetch the requested data it needs to do some work, which additionally increases CPU consumption.
+Server-side caching is an important aspect of `JITServer` and one of the main reasons why `JITServer` client consumes less CPU time compared to local compilation. This document will explain why caching is important, what kind of entities we cache, and provide examples of how some caches work.
+
+1. [Importance of caching](#importance-of-caching)
+2. [Types of caching](#types-of-caching)
+3. [Important caches](#important-caches)
+4. [Cache control](#cache-control)
+
+## Importance of caching
+
+When server is compiling a method, it needs to know a lot of information that is located on the client side, e.g. the addresses of RAM classes, GC mode and its parameters, class hierarchy information, interpreter profiling data and much more.
+To obtain this information, server will make a remote call over the network to the client, client will find it, and send it back to the server. During the course of a compilation server will make hundreds of such calls (on average), assuming no caching is done.
+The number of messages per compiltion will scale with the size of method being compiled and optimization level.
+
+This is bad because remote calls have to go through the network, and many data structures cannot be sent directly and require serialization/deserialization.
+These are expensive operations, both in terms of latency and CPU, especially in terms of CPU. If the server makes hundreds of requests to the client just for one compilation, client will spend more CPU time just sending and receiving data over the network than it would take for it to compile a method locally.
```

Comment:
"both in terms of latency and CPU, especially in terms of CPU." --> "both in terms of latency and CPU"  should be enough